### ISSUE TITLE
Writing data immediately after creating an index

### ISSUE THREAD

**Issue by seanbrant** [pilosa/pilosa#899]
For bugs, please provide the following:

### Expected behavior
Using the python-client I would expect to be able to query an index immediately after syncing the schema.

### Actual behavior
Querying the index results in a index does exist error.

### Steps to reproduce the behavior
```client = PilosaClient(...)
schema = client.schema()
index = schema.index('my-index')
frame = index.frame('my-frame')
client.sync_schema(schema)
client.query(frame.setbit(...))
```

### Information about your environment (OS/architecture, CPU, RAM, cluster/solo, configuration, etc.)
latest python-client and server

**Comment by jaffee** [comment]
@seanbrant are you also creating a frame after creating the index? Also, I assume this issue is triggered by SetBit queries?

**Comment by seanbrant** [comment]
Correct a better example:
```
schema = client.schema()
index = client.index('my-index')
frame = index.frame('my-frame')
schema.sync_schema(schema)
client.query(frame.setbit(...))
```

**Comment by yuce** [comment]
This is not python client specific; it takes about 10 seconds to propagate index creation (tested with the latest release 0.7.1) on a 2-node cluster:
1st node: `pilosa server -b :10501 -d n1_data --cluster.hosts localhost:10501,localhost:10502`
2nd node: `pilosa server -b :10502 -d n2_data --cluster.hosts localhost:10501,localhost:10502 --gossip-port 16000`

After those I've run the following:
```
$ curl http://localhost:10501/status
> {"status":{"Nodes":[{"Host":"localhost:10501","State":"UP"},{"Host":"localhost:10502","State":"UP"}]}}

$ curl http://localhost:10502/status
> {"status":{"Nodes":[{"Host":"localhost:10501","State":"UP"},{"Host":"localhost:10502","State":"UP"}]}}

$ curl http://localhost:10501/index/i1 -d ''
> {}

$ curl http://localhost:10501/status
> {"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP"}]}}

$ curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP"},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}

$ curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP"},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}

$ curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
```

**Comment by yuce** [comment]
This seems to work on the latest server (v0.7.1) and python client (v0.7.0):
```python
import pilosa
client = pilosa.Client(":10501")
schema = client.schema()
index = schema.index('my-index2')
frame = index.frame('my-frame3')
client.sync_schema(schema)
client.query(frame.setbit(1, 100)).result
print(client.query(frame.bitmap(1)).result.bitmap.bits)
```

Prints: `[100L]`

**Comment by seanbrant** [comment]
We have a 6 node cluster. Is it possible that the setbit is being called on a cluster that is not synced?

**Comment by travisturner** [comment]
This (10 second delay) is not expected behavior. Index creation is propagated at time of creation, and should raise an error if any node fails to receive the event. Let's treat this as a bug until we determine what's causing it.

**Comment by seanbrant** [comment]
Looking closer at the logs I might be mistaken. I think it's a frame does not exist error.

**Comment by seanbrant** [comment]
`err=\n\x0fframe not found`

**Comment by yuce** [comment]
@seanbrant are you using single address to create the client, like `pilosa.Client(":10501")` or using a cluster, like: `cluster = pilosa.Cluster(pilosa.URI.address(":10501"), pilosa.URI.address(":10502")...); client = pilosa.Client(cluster)` ?

We used to have some round robin logic in the `Cluster` class which would return a different host for each call, but it was removed and the first healthy host is being returned now.

**Comment by seanbrant** [comment]
We are using a `pilosa.Cluster` object.

**Comment by jaffee** [comment]
@seanbrant I think you need to change the second line in your example to `index = schema.index('my-index')` instead of `index = client.index('my-index')`

**Comment by seanbrant** [comment]
Good catch, we are doing `schema.index` my example was wrong.

**Comment by jaffee** [comment]
ok, so this is still a bug then - I updated the example in the original post and deleted our first two comments @seanbrant

**Comment by travisturner** [comment]
@yuce when I run the example from your first comment in this post, I don't get the same results that you did:
```
~# curl http://localhost:10501/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
~# curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
~# curl http://localhost:10501/index/i1 -d ''
index already exists
~# curl http://localhost:10501/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
[mac]-travis:~# curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
~# curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
~# curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
~# curl http://localhost:10502/status
{"status":{"Nodes":[{"Host":"localhost:10501","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"},"Slices":[0]}]},{"Host":"localhost:10502","State":"UP","Indexes":[{"Name":"i1","Meta":{"ColumnLabel":"columnID"}}]}]}}
```
I am running two nodes locally with `v0.7.1`. Can you provide any more detail on how you got the results that you were seeing?

**Comment by travisturner** [comment]
well, I just noticed the "index already exists". sorry, let me try again.

**Comment by travisturner** [comment]
My best guess as to what is happening here is that the `SendSync()` of the `CreateIndexMessage` in `gossip/gossip.go` failed on one or more nodes in the cluster. A failure in that case returns an error to the node that received the initial `CreateIndex` request.

`handlePostFrame()` in `handler.go` receives the error in that case, but currently it's only sending the error to the logger, and the response back to the client is an empty json object no matter what.

I think we need to always send an error back in the response so that the client is able to handle it appropriately. In this case, the python client should be aware that it needs to raise an exception (or try again) on the failure to create the index property.

**Comment by travisturner** [comment]
I'm not sure if it addresses the problem that you experienced @seanbrant, but in the [cluster-resize](https://github.com/pilosa/pilosa/tree/cluster-resize) branch we have refactored the synchronous communication between nodes to avoid the `SendSync()` possibility that I described below. That work will be released once that branch has been extensively tested.

**Comment by travisturner** [comment]
This issue has been addressed. The `api.CreateIndex()` method now issues a synchronous message to all nodes before returning. If any message fails to be delivered or handled correctly, an error is returned to the client.
